### PR TITLE
Use snapshot's .d file as source inputs in Gradle build.

### LINKED ISSUE #8315: flutter run doesn't pick up code changes in dependent packages with gradle
## Steps to Reproduce (on Linux)

1. `mkdir mylib`
2. `mkdir mylib/lib`
3. `echo "name: mylib" > mylib/pubspec.yaml`
4. `echo "const int generation = 1;" >> mylib/lib/bar.dart`
5. `flutter create foobar`
6. `cd foobar`
7. Add the following to the "dependencies" section of your pubspec.yaml:

```yaml
  mylib:
    path: ../mylib
```

8. Add an `import "package:mylib/bar.dart";` to `lib/main.dart`
9. Change `'Button tapped $_counter time${ _counter == 1 ? '' : 's' }.',` to `'$generation',`
10. `flutter run`

Notice that you see the number 1 in the middle of the screen.

11. Edit `../mylib/lib/bar.dart` to change `generation` from `1` to `2`.
12. `flutter run`

Notice that you still see the number 1 in the middle of the screen instead of 2, which is what the code currently does.

## Discussion

As part of the build, we output `./android/app/build/intermediates/flutter/debug/snapshot_blob.bin.d`, which lists the Dart files that were read during the build.  If any of those files change, we need to recreate the snapshot to see the code change.  It appears the gradle build is ignoring this file and only considering code changes inside the gradle project itself.

/cc @jakobr-google @eseidelGoogle @goderbauer

### LINKED ISSUE #8607: Gradle build fails after engine update
## Steps to Reproduce

Running on flutter from 3 days ago app worked.  After a flutter upgrade this morning (3/7) the app won't run.  Running from Android Studio.

## Logs
```
03-07 07:26:15.675 12331-12331/com.askllc.vinteresting E/ApkAssets: Error while loading asset assets/icudtl.dat: java.io.FileNotFoundException: This file can not be opened as a file descriptor; it is probably compressed
03-07 07:26:15.698 12331-12331/com.askllc.vinteresting I/Adreno: QUALCOMM build                   : e781517, I9267f384df
                                                                 Build Date                       : 10/26/16
                                                                 OpenGL ES Shader Compiler Version: XE031.09.00.03
                                                                 Local Branch                     : 
                                                                 Remote Branch                    : refs/tags/AU_LINUX_ANDROID_LA.UM.5.5.R1.07.00.00.269.009
                                                                 Remote Branch                    : NONE
                                                                 Reconstruct Branch               : NOTHING
03-07 07:26:15.802 12331-12375/com.askllc.vinteresting E/flutter: [ERROR:../../lib/tonic/logging/dart_error.cc(16)] Wrong script snapshot version, expected '795c4aba89b54b4c51894de889493cdf' found '850b51cd3b18a02dc9f31c0b233f3fe3'
03-07 07:26:15.832 12331-12375/com.askllc.vinteresting E/flutter: [ERROR:../../lib/tonic/logging/dart_error.cc(16)] Unhandled exception:
                                                                  NoSuchMethodError: No top-level getter 'main' declared.
                                                                  Receiver: top-level
                                                                  Tried calling: main
                                                                  #0      NoSuchMethodError._throwNew (dart:core-patch/errors_patch.dart:194)
                                                                  #1      _getMainClosure (file:///b/build/slave/Linux_Engine/build/src/flutter/lib/ui/natives.dart:55)
03-07 07:26:15.916 12331-12378/com.askllc.vinteresting D/libc: [NET] android_getaddrinfofornetcontext+,hn 9(0x3132372e302e30),sn(),hints(known),family 0,flags 1024, proc=com.askllc.vinteresting
03-07 07:26:15.922 12331-12378/com.askllc.vinteresting D/libc: [NET] android_getaddrinfofornetcontext+,hn 9(0x3132372e302e30),sn(),hints(known),family 0,flags 1024, proc=com.askllc.vinteresting
03-07 07:26:16.009 12331-12390/com.askllc.vinteresting I/flutter: Diagnostic server listening on http://127.0.0.1:35175/
03-07 07:26:16.020 12331-12390/com.askllc.vinteresting I/flutter: Observatory listening on http://127.0.0.1:33483/
```
## Flutter Doctor

Paste the output of running `flutter doctor` here.

[√] Flutter (on Windows, channel master)
    • Flutter at C:\dev\flutter
    • Framework revision 24eeddc053 (10 hours ago),
    • Engine revision f9c1f5fa53
    • Tools Dart version 1.23.0-dev.2.0

[√] Android toolchain - develop for Android devices (Android SDK 25.0.2)
    • Android SDK at C:\Users\Rob\AppData\Local\Android\sdk
    • Platform android-25, build-tools 25.0.2
    • Java(TM) SE Runtime Environment (build 1.8.0_121-b13)

[√] Android Studio (version 2.3)
    • Android Studio at C:\Program Files\Android\Android Studio
    • Gradle version 3.2

[√] IntelliJ IDEA Community Edition (version 2016.3)
    • Dart plugin version 163.13137
    • Flutter plugin version 0.1.11.1

[√] Connected devices
    • 2PS64 • HT69N0100346 • android-arm • Android 7.0 (API 24)

### LINKED ISSUE #8687: gradle build not noticing that framework changed
* Delete Flutter Gallery from your phone.
* `cd .../examples/flutter_gallery`
* `flutter build clean` (once https://github.com/flutter/flutter/issues/8686 is fixed; if it's not fixed, delete `android/build` and `android/app/build` manually)
* `flutter run`
* Edit a framework file. For example, add a syntax error to `.../packages/flutter/lib/src/widgets/scroll_controller.dart`.
* Verify (by hot reloading, `r`) that the syntax error is there.
* `^C`
* `flutter run`
* Notice that the framework change has not taken effect.

cc @jakobr-google @mit-mit 

This is rather painful for us framework developers. :-)

### PR DESCRIPTION
**Pull request by jakobr-google**
If we don't yet have a .d file (first build), fall back to using the
.dart files in the current directory. This enables us to detect changes
in dependent source files (Flutter framework, packages outside the
source directory, etc.), and re-generate the snapshots as needed.

Unfortunately, Gradle requires knowing the source files before executing
the task, and can't update them after building, so Gradle considers the
second build to be out-of-date (because it has more input files than the
first build). Sub-sequent builds have the correct dependency
information, and will be skipped if the source files haven't changed.

Also added a dependency on gen_snapshot. The snapshot ABI isn't stable,
so we need to re-generate the snapshots when we roll the Dart SDK
dependency.

Fixes #8315
Fixes #8687
Fixes #8607

### CODE REVIEW

File: packages/flutter_tools/gradle/flutter.gradle [hunk 1]
```diff
@@ -207,9 +207,42 @@ class FlutterTask extends DefaultTask {
...
     FileCollection getSourceFiles() {
-        return project.fileTree(dir: sourceDir, exclude: ['android', 'ios'], include: ['**/*.dart', 'pubspec.yaml', 'flutter.yaml'])
+        File dependenciesFile
+        if (buildMode != 'debug') {
+            dependenciesFile = project.file("${intermediateDir}/snapshot.d")
+        } else {
+            dependenciesFile = project.file("${intermediateDir}/snapshot_blob.bin.d")
+        }
+        FileCollection sources = readDependencies(dependenciesFile)
+        if (sources != null) {
```
Comment (mravn-google) [inline packages/flutter_tools/gradle/flutter.gradle]: Please add a comment describing the situations `sources == null` and `sources !=null`

File: packages/flutter_tools/gradle/flutter.gradle [hunk 2]
```diff
@@ -207,9 +207,42 @@ class FlutterTask extends DefaultTask {
         }
     }
 
+    FileCollection readDependencies(File dependenciesFile) {
+        if (dependenciesFile.exists()) {
+            try {
+                // Dependencies file has Makefile syntax:
+                //   <target> <files>: <source> <files> <separated> <by> <space>
+                String depText = dependenciesFile.text
```
Comment (mravn-google) [inline packages/flutter_tools/gradle/flutter.gradle]: Can a dependency files have comments? Like lines of the form
```# TODO(mravn): remove dependencies on bigwig```

**Review by mravn-google — APPROVED** [review]
LGTM

File: packages/flutter_tools/gradle/flutter.gradle (hunk 2 above)
Comment (jakobr-google) [inline packages/flutter_tools/gradle/flutter.gradle]: It's technically in 'Makefile' format, which does allow comments. However, gen_snapshot only emits one line containing 'target: sources'.

File: packages/flutter_tools/gradle/flutter.gradle (hunk 1 above)
Comment (jakobr-google) [inline packages/flutter_tools/gradle/flutter.gradle]: Done.

**Comment by eseidelGoogle** [comment]
FYI @jason-simmons @abarth

**Pull request merged**